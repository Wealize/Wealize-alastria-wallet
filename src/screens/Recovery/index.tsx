import React, { useEffect, useState } from 'react'
import FlashMessage from 'react-native-flash-message'
import { NavigationFunctionComponent } from 'react-native-navigation'
import Snackbar from 'react-native-snackbar'

import TouchableButton from '../../components/TouchableButton'
import { GET_EXTERNAL_SCREENS_OPTIONS } from '../../constants/options'
import { RECOVERY } from '../../constants/text'
import { checkSecurityPhrase, changePin } from '../../utils/keychain'
import { NavigationProps, popScreen } from '../../utils/navigation-utils'
import {
  MainContainer,
  Title,
  BackButton,
  Subtitle,
  InputsContainer,
  InputStyled,
  ContainerBottom,
  PhraseInput
} from './styles'

const Recovery: NavigationFunctionComponent = ({
  componentId
}: NavigationProps) => {
  const [securityPhrase, setSecurityPhrase] = useState('')
  const [firstPin, setFirstPin] = useState('')
  const [secondPin, setSecondPin] = useState('')
  const [isPinFilled, setIsPinFilled] = useState(false)
  const [isLoading, setIsLoading] = useState(false)

  const flashMessage: React.RefObject<any> = React.createRef()

  useEffect(() => {
    setIsPinFilled(arePinEquals())
  }, [firstPin, secondPin])

  const setNewPin = async (securityPhrase: string, pin: string) => {
    setIsLoading(true)

    const isCorrectPhrase = await checkSecurityPhrase(securityPhrase)
    isCorrectPhrase ? changePinAndClearForm(pin) : showError()
  }

  const changePinAndClearForm = async (pin: string) => {
    const isChanged = await changePin(pin)
    if (isChanged) {
      clearForm()
      Snackbar.show({
        text: RECOVERY.SUCCESS,
        duration: Snackbar.LENGTH_SHORT
      })
      setIsLoading(false)
    }
  }

  const clearForm = (): void => {
    setSecurityPhrase('')
    setFirstPin('')
    setSecondPin('')
    setIsLoading(false)
  }

  const renderErrorMessage = () => {
    if (firstPin === '' || secondPin === '' || arePinEquals()) {
      return ''
    }
    return RECOVERY.ERROR_PIN
  }

  const arePinEquals = () => {
    return firstPin !== '' && secondPin !== '' && firstPin === secondPin
  }

  const showError = (): void => {
    Snackbar.show({
      text: RECOVERY.ERROR_PHRASE,
      duration: Snackbar.LENGTH_SHORT
    })
    setIsLoading(false)
  }

  return (
    <MainContainer>
      <BackButton
        type="clear"
        title={RECOVERY.BACK}
        onPress={() => popScreen(componentId)}
      />
      <Title>{RECOVERY.TITLE}</Title>
      <Subtitle>{RECOVERY.SUBTITLE}</Subtitle>

      <InputsContainer>
        <PhraseInput
          value={securityPhrase}
          multiline
          placeholder={RECOVERY.FORM_PLACEHOLDER.SECURITY_KEY}
          onChangeText={(value) => setSecurityPhrase(value)}
        />
        <InputStyled
          value={firstPin}
          keyboardType="decimal-pad"
          onChangeText={(value) => setFirstPin(value.replace(/[^0-9]/g, ''))}
          placeholder={RECOVERY.FORM_PLACEHOLDER.PIN}
        />
        <InputStyled
          value={secondPin}
          keyboardType="decimal-pad"
          onChangeText={(value) => setSecondPin(value.replace(/[^0-9]/g, ''))}
          placeholder={RECOVERY.FORM_PLACEHOLDER.REPIN}
          errorMessage={renderErrorMessage()}
        />
      </InputsContainer>

      <ContainerBottom>
        <TouchableButton
          title={isLoading ? RECOVERY.LOADING : RECOVERY.BUTTON}
          disabled={!isPinFilled || isLoading}
          onPress={() => setNewPin(securityPhrase, secondPin)}
        />
      </ContainerBottom>
      <FlashMessage ref={flashMessage} position="top" />
    </MainContainer>
  )
}

Recovery.options = {
  ...GET_EXTERNAL_SCREENS_OPTIONS,
  statusBar: { ...GET_EXTERNAL_SCREENS_OPTIONS.statusBar, style: 'dark' }
}

export default Recovery
