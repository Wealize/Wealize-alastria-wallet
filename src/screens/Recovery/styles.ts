import { Text, View, TextInput } from 'react-native'
import styled from 'styled-components'
import { Button, Input } from 'react-native-elements'

import ViewContainer from '../../components/ViewContainer'
import { Colors } from '../../utils/themes'

export const MainContainer = styled(ViewContainer)`
  background-color: #ffffff;
  padding-left: 0;
  padding-right: 0;
`

export const BackButton = styled(Button).attrs({
  containerStyle: {
    marginTop: 25,
    alignSelf: 'flex-start'
  },
  buttonStyle: {
    justifyContent: 'flex-start',
    marginLeft: 18
  },
  titleStyle: {
    color: Colors.activeInput,
    fontFamily: 'SourceSansPro',
    fontSize: 14
  },
  icon: {
    name: 'arrow-back-ios',
    size: 20,
    color: Colors.activeInput
  }
})``

export const Title = styled(Text)`
  font-size: 20px;
  font-weight: bold;
  padding-top: 5%;
  margin-left: 9%;
`

export const Subtitle = styled(Text)`
  margin-top: 13%;
  margin-bottom: 10%;
  margin-left: 9%;
  margin-right: 7%;
  font-size: 13px;
  text-align: justify;
`

export const InputsContainer = styled(View)`
  margin-left: 6%;
  margin-right: 8%;
  margin-bottom: 6%;
`

export const PhraseInput = styled(TextInput).attrs({
  selectionColor: Colors.activeInput,
  placeholderTextColor: 'rgba(60, 60, 67, 0.29)'
})`
  height: 90px;
  border-radius: 20px;
  border-width: 1px;
  border-color: rgba(60, 60, 67, 0.29);
  text-align: center;
  font-size: 16px;
  margin-bottom: 5%;
`

export const InputStyled = styled(Input).attrs({
  selectionColor: Colors.activeInput,
  errorStyle: {
    color: Colors.error
  }
})``

export const ContainerBottom = styled(View)`
  flex: 1;
  justify-content: flex-end;
  align-items: center;
  margin: 20px auto 40px auto;
  width: 100%;
`