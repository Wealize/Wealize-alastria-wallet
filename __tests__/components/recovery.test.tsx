import React from 'react'
import renderer from 'react-test-renderer'

import Recovery from '../../src/screens/Recovery'

describe('OnboardginComponent', () => {
  it('should render correctly', () => {
    const result = renderer.create(<Recovery componentId={'1'} />)
    expect(result).toMatchSnapshot()
  })
})
