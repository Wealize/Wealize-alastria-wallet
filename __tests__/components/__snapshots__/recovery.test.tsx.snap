// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardginComponent should render correctly 1`] = `
<View
  className="sc-dlfnuX sc-hKgJUU euTbpO iRiBIE"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        Object {
          "alignSelf": "flex-start",
          "marginTop": 25,
        },
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginLeft": 18,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "marginHorizontal": 5,
              "overflow": "hidden",
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#00b0ff",
                      "fontSize": 20,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#00b0ff",
              "fontFamily": "SourceSansPro",
              "fontSize": 14,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Volver
        </Text>
      </View>
    </View>
  </View>
  <Text
    className="sc-jSgvzq cHdwbB"
  >
    Recuperar acceso
  </Text>
  <Text
    className="sc-gKseQn gNShDS"
  >
    Para volver a acceder a tu aplicación debes introduccir tu frase de seguridad y crear un pin de acceso nuevo. 
  </Text>
  <View
    className="sc-iBPTik gURcWP"
  >
    <TextInput
      allowFontScaling={true}
      className="sc-fubCzh bKmelu"
      multiline={true}
      onChangeText={[Function]}
      placeholder="Frase de Seguridad"
      placeholderTextColor="rgba(60, 60, 67, 0.29)"
      rejectResponderTermination={true}
      selectionColor="#00b0ff"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          className="sc-pGacB"
          editable={true}
          keyboardType="decimal-pad"
          onChangeText={[Function]}
          placeholder="Nuevo Pin"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          selectionColor="#00b0ff"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#EE3F61",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          className="sc-pGacB"
          editable={true}
          keyboardType="decimal-pad"
          onChangeText={[Function]}
          placeholder="Repetir Pin"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          selectionColor="#00b0ff"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#EE3F61",
            "fontSize": 12,
            "margin": 5,
          }
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    className="sc-jrAFXE bGAxgA"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#CCCCCC",
            "borderRadius": 45,
            "height": 39,
            "justifyContent": "center",
            "width": 224,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 16,
            }
          }
        >
          Recuperar
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": -48,
          },
        ],
        "zIndex": 99,
      }
    }
  />
</View>
`;
